# Problem Statement


New York City is the most densely populated city in the United States of America and has varying dining options. Based on 2019 information, there are approximately 8.6 million people within  the city.  Based on research performed by CBRE, New Yorkers spend on average USD8,082 per year compared to their peers across the United States that spend USD 3,512  per year. Based on this statistics, the competition is fierce within the area. In New York you find there are clusters of food areas that focus on certain type of cuisine such as Italian, Spanish, Korean and Chinese. These districts  provide an opportunity for people to experience different cultures and cuisine.  


There is no formal African district where the rich African culture can be explored and showcased to the broader public. The aim of this analysis will be to understand whether there are a number of African restaurants clustered in certain areas and whether there is an opportunity to rebrand the area to an African town to incentivise patrons to visit the area.  As part of this analysis, African restaurants will be visualised in the New York City area.


# Data

For  this analysis the following data will be required:
   - New York City Data
   - Foursquare API Data; and
   - Geospace data

The sources of information are unpacked below in more detail. 
            

###### New York City Data
New York City data that contains a list of the various neighbourhoods, which includes longitude and latitude
Data Source: https://cocl.us/new_york_dataset
Description: This data set contains the required  data to explore various neighbourhoods of New York City.

###### African restuarants in each neighbourhood of New York City
 Data source: Foursquare API
 Description: This API allows the code to get all the venues in each neighbourhood. This will be filtered based on the African restaurant venues.



###### Geospace data
Source: https://data.cityofnewyork.us/City-Gorvenmetn/Burrough-Boundaries/tgmj-j8zm
Description: By using the geo space data we will get the New York Borough boundaries that will assist in visualizing a choropleth map. 

###### Approach

###### Collect the new york city data from https://cocl.us/new_york_dataset

Using FourSquare API we will find all venues for each neighborhood.

Filter out all venues that are African Resturants.

This will be visualised and the higher areas will be 

Visualize the Ranking of neighborhoods using folium library(python)

###### Questions that can be asked using the above dataset

Where ware the African restuarants situated in New York?

Which Buroughs has the most number of African Restuarants?

Which areas lack African restaurants?


###### Analysis

We will import the required libraries for python.

    - pandas and numpy for handling data.
    - request module for using FourSquare API.
    - geopy to get co-ordinates of City of New York.
    - folium to visualize the results on a map

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from pprint import pprint # data pretty printer

import requests # library to handle requests
from bs4 import BeautifulSoup  # library to handle web scraping

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

!pip install folium # map rendering library

import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors # Matplotlib and associated plotting modules

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

from collections import Counter # count occurrences 
print ('Libraries imported')

     |████████████████████████████████| 102kB 6.7MB/s ta 0:00:011
Libraries imported


Next step will be to extract venue information from Foursquare using the Foursquare API

In [4]:
import wget
import os

if os.path.exists('newyork_data.json'):
    os.remove('newyork_data.json')

wget.download('https://cocl.us/new_york_dataset/newyork_data.json')

print('\nData downloaded!')


Data downloaded!


In [5]:
# open & load downloaded file, print its type
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)
type(newyork_data)

dict

In [6]:
print ("newyork_data key and value's length are: ") 
for key, value in newyork_data.items(): 
    print (key, len(str(value)))

newyork_data key and value's length are: 
type 17
totalFeatures 3
features 124743
crs 70
bbox 74


In [7]:
neighborhoods_data = newyork_data['features']
pprint(neighborhoods_data[305])

{'geometry': {'coordinates': [-74.08173992211962, 40.61731079252983],
              'type': 'Point'},
 'geometry_name': 'geom',
 'id': 'nyu_2451_34572.306',
 'properties': {'annoangle': 0.0,
                'annoline1': 'Fox',
                'annoline2': 'Hills',
                'annoline3': None,
                'bbox': [-74.08173992211962,
                         40.61731079252983,
                         -74.08173992211962,
                         40.61731079252983],
                'borough': 'Staten Island',
                'name': 'Fox Hills',
                'stacked': 2},
 'type': 'Feature'}


###### Transform data into a pands dataframe

In [8]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then lets loop through the data and fill the dataframe one row at a time

In [9]:
for data in neighborhoods_data:
    borough = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [9]:
neighborhoods.head()


Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [10]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


Use geopy library to get the latitude and longitude values of New York City

In [11]:
address = 'New York City, NY'
location = None

# define an instance of the geocoder -> ny_explorer
while location == None:
    try:
        geolocator = Nominatim(user_agent="ny_explorer")
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
    except:
        pass
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


Create a map of New York with neighborhoods superimposed on top.

In [13]:
import folium
!pip install folium # map rendering library
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

###### Define Foursquare Credentials and Version

Utilizing the Foursquare API to explore the neighborhoods

In [14]:
CLIENT_ID = 'VTEXLDCRMNRHSALNG3N2A1FFLRNK3ON03M0PYJYU4RD1CS1Z' # your Foursquare ID
CLIENT_SECRET = '0UK31K23U0UZ4HGEMKP44VQXB1QJKKXL1LKXWEC5YDP40HOS' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VTEXLDCRMNRHSALNG3N2A1FFLRNK3ON03M0PYJYU4RD1CS1Z
CLIENT_SECRET:0UK31K23U0UZ4HGEMKP44VQXB1QJKKXL1LKXWEC5YDP40HOS


In [15]:
url = 'https://api.foursquare.com/v2/venues/categories?&client_id={}&client_secret={}&v={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION)
category_results = requests.get(url).json()

In [16]:
for key, value in category_results['response']['categories'][0].items():
    print(key, len(str(value)))

id 24
name 20
pluralName 20
shortName 20
icon 98
categories 15910


In [17]:
category_list = category_results['response']['categories']

In [18]:
len(category_list)

10

In [17]:
for data in category_list:
    print(data['id'], data['name'])

4d4b7104d754a06370d81259 Arts & Entertainment
4d4b7105d754a06372d81259 College & University
4d4b7105d754a06373d81259 Event
4d4b7105d754a06374d81259 Food
4d4b7105d754a06376d81259 Nightlife Spot
4d4b7105d754a06377d81259 Outdoors & Recreation
4d4b7105d754a06375d81259 Professional & Other Places
4e67e38e036454776db1fb3a Residence
4d4b7105d754a06378d81259 Shop & Service
4d4b7105d754a06379d81259 Travel & Transport


In [20]:
# function to flatten a 'parent_id' category, returns all categories if checkParentID = False
def flatten_Hierarchy(category_list, checkParentID, category_dict, parent_id = ''):
    for data in category_list:
        
        if checkParentID == True and data['id'] == parent_id:
            category_dict[data['id']] = data['name']
            flatten_Hierarchy(category_list = data['categories'], checkParentID = False, category_dict = category_dict)
        
        elif checkParentID == False:
            category_dict[data['id']] = data['name']
            if len(data['categories']) != 0:
                flatten_Hierarchy(category_list = data['categories'], checkParentID = False, category_dict = category_dict)
    
    return category_dict


In [21]:
# this function takes the parent ID and returns 'name' and 'id' of all the sub-categories
category_dict = flatten_Hierarchy(category_list, checkParentID=True, category_dict = {}, parent_id = '4d4b7105d754a06374d81259')

Explore the first neighborhood to understand the results of GET Request

Get the neighborhood's name.

In [22]:
neighborhoods.loc[0, 'Neighborhood']


'Wakefield'

Get the neighborhood's latitude and longitude values.

In [23]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Wakefield are 40.89470517661, -73.84720052054902.


Now, let's get the Food that is in Wakefield within a radius of 500 meters.


In [24]:
LIMIT = 1 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
categoryId = '4d4b7105d754a06374d81259' # category ID for "Food"

# create URL

url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&categoryId={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius,
    categoryId,
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/search?&client_id=VTEXLDCRMNRHSALNG3N2A1FFLRNK3ON03M0PYJYU4RD1CS1Z&client_secret=0UK31K23U0UZ4HGEMKP44VQXB1QJKKXL1LKXWEC5YDP40HOS&v=20180605&ll=40.89470517661,-73.84720052054902&radius=500&categoryId=4d4b7105d754a06374d81259&limit=1'

In [25]:
results = requests.get(url).json()

In [24]:
results['response']['venues']

[{'id': '4eb046abb8f7116f21081aeb',
  'name': 'Baychester Avenue Food Truck',
  'location': {'address': 'Baychester Avenue',
   'crossStreet': '233rd Street',
   'lat': 40.890479268747484,
   'lng': -73.84272476029565,
   'labeledLatLngs': [{'label': 'display',
     'lat': 40.890479268747484,
     'lng': -73.84272476029565}],
   'distance': 602,
   'postalCode': '10466',
   'cc': 'US',
   'city': 'Bronx',
   'state': 'NY',
   'country': 'United States',
   'formattedAddress': ['Baychester Avenue (233rd Street)',
    'Bronx, NY 10466',
    'United States']},
  'categories': [{'id': '4bf58dd8d48988d1cb941735',
    'name': 'Food Truck',
    'pluralName': 'Food Trucks',
    'shortName': 'Food Truck',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/streetfood_',
     'suffix': '.png'},
    'primary': True}],
  'referralId': 'v-1590950017',
  'hasPerk': False}]

The category name of the venue 'Carvel Ice Cream' is 'Food'.

As, our aim is to segment the neighborhoods of NYC with respect to the Food in its vicinity. We need to proceed further to fetch this data from all the 306 neighborhoods' venues.

###### Let's create a function to repeat the following process to all the neighborhoods in NYC:

Loop through neighborhoods
    -  Create the API request URL with radius=500, LIMIT=100
    - Make the GET request
    - For each neighborhood, return only relevant information for each nearby venue
Append all nearby venues to a list
Unfold the list & append it to dataframe being returned
The categoryId parameter in the API request URL can be a comma seperated string. So, lets create a comma seperated string from category_dict.

In [26]:
categoryId_list = []
for key, value in category_dict.items():
    categoryId_list.append(key)
categoryId = ','.join(categoryId_list)

The above process turns out to be redundant because if a top-level category is specified in the GET Request, all sub-categories will also match the query. But it is an interesting way to retrieve all the sub-categories ID with name.

In [27]:
def getNearbyFood(names, latitudes, longitudes, radius=1000, LIMIT=500):
    not_found = 0
    print('***Start ', end='')
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(' .', end='')
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&categoryId={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            "4d4b7105d754a06374d81259", # "Food" category id
            LIMIT)
            
        try:
            # make the GET request
            results = requests.get(url).json()['response']['venues']
            
            # return only relevant information for each nearby venue
            venues_list.append([(
                name, 
                lat, 
                lng, 
                v['name'], 
                v['location']['lat'], 
                v['location']['lng'],  
                v['categories'][0]['name']) for v in results])
        except:
            not_found += 1


    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    print("\nDone*** with {} venues with incompelete information.".format(not_found))
    return(nearby_venues)


Let's use pickle library to serialize the information retrieved from GET requests. This step will counter any redundant requests to the Foursquare API.

In [28]:
import pickle # to serialize and deserialize a Python object structure
try:
    with open('nyc_food_venues.pkl', 'rb') as f:
        nyc_venues = pickle.load(f)
    print("---Dataframe Existed and Deserialized---")
except:
    nyc_venues = getNearbyFood(names=neighborhoods['Neighborhood'],
                                        latitudes=neighborhoods['Latitude'],
                                        longitudes=neighborhoods['Longitude']
                                       )
    with open('nyc_food_venues.pkl', 'wb') as f:
        pickle.dump(nyc_venues, f)
    print("---Dataframe Created and Serialized---")

***Start  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
Done*** with 0 venues with incompelete information.
---Dataframe Created and Serialized---


Let's check the size of the resulting dataframe

In [29]:
print(nyc_venues.shape)
nyc_venues.head()

(13766, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                                  Dunkin'       40.890459       -73.849089   
1                         Carvel Ice Cream       40.890487       -73.848568   
2  Cooler Runnings Jamaican Restaurant Inc       40.898083       -73.850259   
3            Ming's Wok Chinese Restaurant       40.893542       -73.858689   
4                              Burger King       40.895540       -73.856460   

         Venue Category  
0            Donut Shop  
1        Ice Cream Shop  
2  Caribbean Restaurant  
3    Chinese Restaurant  
4  Fast Food Restaurant

Let's find out how many unique categories can be curated from all the returned venues

In [30]:
print('There are {} uniques categories.'.format(len(nyc_venues['Venue Category'].unique())))
nyc_venues.groupby('Venue Category')['Venue Category'].count().sort_values(ascending=False)

There are 193 uniques categories.


Venue Category
Deli / Bodega                               1313
Pizza Place                                 1161
Coffee Shop                                  846
Chinese Restaurant                           645
Fast Food Restaurant                         639
Bakery                                       599
Donut Shop                                   599
Italian Restaurant                           476
Bagel Shop                                   404
Caribbean Restaurant                         374
Fried Chicken Joint                          347
Mexican Restaurant                           339
Café                                         324
Sandwich Place                               320
American Restaurant                          318
Ice Cream Shop                               296
Diner                                        249
Food                                         239
Restaurant                                   220
Burger Joint                                 195
Food 

In [29]:
nyc_venues.groupby('Venue Category')['Venue Category'].count().sort_values(ascending=False)

Venue Category
Deli / Bodega                               1312
Pizza Place                                 1162
Coffee Shop                                  846
Chinese Restaurant                           643
Fast Food Restaurant                         642
Bakery                                       600
Donut Shop                                   600
Italian Restaurant                           478
Bagel Shop                                   405
Caribbean Restaurant                         374
Fried Chicken Joint                          348
Mexican Restaurant                           340
Sandwich Place                               319
American Restaurant                          319
Café                                         319
Ice Cream Shop                               294
Diner                                        249
Food                                         239
Restaurant                                   219
Burger Joint                                 196
Food 

As we are interested in exploring the diversity of the neighborhood. Let's remove the generalized categories, like Coffee Shop, Cafe, etc.

In [30]:
# list all the categories
unique_categories = nyc_venues['Venue Category'].unique().tolist()
print(', '.join(str(x) for x in unique_categories))


Donut Shop, Ice Cream Shop, Caribbean Restaurant, Fast Food Restaurant, Chinese Restaurant, Bakery, Deli / Bodega, Fried Chicken Joint, Thai Restaurant, Diner, Pizza Place, Latin American Restaurant, Southern / Soul Food Restaurant, Coffee Shop, American Restaurant, Food, Food Truck, Bagel Shop, Mexican Restaurant, Seafood Restaurant, Dumpling Restaurant, BBQ Joint, Italian Restaurant, Restaurant, Sandwich Place, Spanish Restaurant, Juice Bar, Grocery Store, Wings Joint, Café, Burger Joint, Breakfast Spot, Asian Restaurant, Buffet, Sushi Restaurant, Tapas Restaurant, Steakhouse, Bar, Japanese Restaurant, German Restaurant, Mediterranean Restaurant, Cuban Restaurant, Lounge, Pub, Indian Restaurant, Fish & Chips Shop, Cafeteria, Taco Place, Vegetarian / Vegan Restaurant, Empanada Restaurant, Salvadoran Restaurant, Other Nightlife, Soup Place, Arcade, Food Court, New American Restaurant, Dessert Shop, Modern European Restaurant, Frozen Yogurt Shop, French Restaurant, Convenience Store, Su

In [31]:
# manually create a list of generalized categories
general_categories = ['Dessert Shop','Food','Ice Cream Shop','Donut Shop','Bakery','Sandwich Place','Comfort Food Restaurant',
                    'Deli / Bodega','Food Truck','Bagel Shop','Burger Joint','Restaurant','Frozen Yogurt Shop','Coffee Shop',
                    'Diner','Wings Joint','Café','Juice Bar','Breakfast Spot','Grocery Store','Bar','Cupcake Shop',
                    'Pub','Fish & Chips Shop','Cafeteria','Other Nightlife','Arcade','Hot Dog Joint','Food Court',
                    'Health Food Store','Convenience Store','Food & Drink Shop','Cocktail Bar','Cheese Shop',
                    'Snack Place','Sports Bar','Lounge','Theme Restaurant','Buffet','Bubble Tea Shop','Building',
                    'Irish Pub','College Cafeteria','Tea Room','Supermarket','Hotpot Restaurant','Gastropub','Beer Garden',
                    'Fish Market','Beer Bar','Clothing Store','Music Venue','Bistro','Salad Place','Wine Bar','Gourmet Shop',
                    'Indie Movie Theater','Art Gallery','Gift Shop','Pie Shop','Fruit & Vegetable Store',
                    'Street Food Gathering','Dive Bar','Factory','Farmers Market','Mac & Cheese Joint','Creperie',
                    'Candy Store','Event Space','Skating Rink','Miscellaneous Shop','Gas Station','Organic Grocery',
                    'Pastry Shop','Club House','Flea Market','Hotel','Furniture / Home Store','Bookstore','Pet Café',
                    'Gym / Fitness Center','Flower Shop','Financial or Legal Service','Hotel Bar','Hookah Bar','Poke Place',
                    'Market','Gluten-free Restaurant','Smoothie Shop','Butcher','Food Stand','Beach Bar','Beach',
                    'Soup Place','Rock Club','Residential Building (Apartment / Condo)','Laundry Service',
                    'Government Building','Bowling Alley','Nightclub','Park','Moving Target']

In [32]:
# fetch all the required food categories
food_categories =  list(set(unique_categories) - set(general_categories))
print(', '.join(str(x) for x in food_categories))

Italian Restaurant, Bike Shop, Japanese Restaurant, Tibetan Restaurant, Mexican Restaurant, Filipino Restaurant, Arepa Restaurant, Indian Restaurant, African Restaurant, South American Restaurant, Souvlaki Shop, Indonesian Restaurant, Beer Store, Varenyky restaurant, American Restaurant, Salvadoran Restaurant, Burrito Place, English Restaurant, Caribbean Restaurant, Asian Restaurant, Spanish Restaurant, Island, Empanada Restaurant, Kosher Restaurant, Brazilian Restaurant, Caucasian Restaurant, Yemeni Restaurant, Portuguese Restaurant, Hawaiian Restaurant, Ramen Restaurant, Sushi Restaurant, Kebab Restaurant, Dumpling Restaurant, Tex-Mex Restaurant, Thai Restaurant, Southern / Soul Food Restaurant, Vietnamese Restaurant, Falafel Restaurant, Himalayan Restaurant, Israeli Restaurant, Dosa Place, Fast Food Restaurant, Seafood Restaurant, Cajun / Creole Restaurant, Modern Greek Restaurant, Noodle House, Pakistani Restaurant, Persian Restaurant, Cantonese Restaurant, Taco Place, Halal Restau

Let's remove all the generalized categories, like Coffee Shop, Cafe, etc.

In [33]:
nyc_venues = nyc_venues[nyc_venues['Venue Category'].isin(food_categories)].reset_index()
nyc_venues.head(5)

index Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      2    Wakefield              40.894705              -73.847201   
1      3    Wakefield              40.894705              -73.847201   
2      5    Wakefield              40.894705              -73.847201   
3      8    Wakefield              40.894705              -73.847201   
4      9    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0  Cooler Runnings Jamaican Restaurant Inc       40.898083       -73.850259   
1                              Burger King       40.895540       -73.856460   
2            Ming's Wok Chinese Restaurant       40.893542       -73.858689   
3                               McDonald's       40.892779       -73.857473   
4                               McDonald's       40.902645       -73.849485   

         Venue Category  
0  Caribbean Restaurant  
1  Fast Food Restaurant  
2    Chinese Restaurant  
3  Fast Food Restaurant  
4  Fast Food Restaurant

In [34]:
print('There are {} unique FOOD categories.'.format(len(nyc_venues['Venue Category'].unique())))

There are 99 unique FOOD categories.


Let's find out how many unique venues can be curated from all the returned venues

In [35]:
print('There are {} uniques venues.'.format(len(nyc_venues['Venue'].unique())))

There are 3193 uniques venues.


In [37]:
# one hot encoding
nyc_onehot = pd.get_dummies(nyc_venues[['Venue Category']], prefix="", prefix_sep="")
nyc_onehot.head()


Afghan Restaurant  African Restaurant  American Restaurant  \
0                  0                   0                    0   
1                  0                   0                    0   
2                  0                   0                    0   
3                  0                   0                    0   
4                  0                   0                    0   

   Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
0                 0                       0                 0   
1                 0                       0                 0   
2                 0                       0                 0   
3                 0                       0                 0   
4                 0                       0                 0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Beer Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Belgian Restaurant  Bike Shop  Brazilian Restaurant  Burrito Place  \
0                   0          0                     0              0   
1                   0          0                     0              0   
2                   0          0                     0              0   
3                   0          0                     0              0   
4                   0          0                     0              0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                          0                     0                     1   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
0                     0                   0                     0   
1                     0                   0                     0   
2                     0                   1                     0   
3                     0                   0                     0   
4                     0                   0                     0   

   Cuban Restaurant  Dim Sum Restaurant  Dosa Place  Dumpling Restaurant  \
0                 0                   0           0                    0   
1                 0                   0           0                    0   
2                 0                   0           0                    0   
3                 0                   0           0                    0   
4                 0                   0           0                    0   

   Eastern European Restaurant  Empanada Restaurant  English Restaurant  \
0                            0                    0                   0   
1                            0                    0                   0   
2                            0                    0                   0   
3                            0                    0                   0   
4                            0                    0                   0   

   Ethiopian Restaurant  Falafel Restaurant  Fast Food Restaurant  \
0                     0                   0                     0   
1                     0                   0                     1   
2                     0                   0                     0   
3                     0                   0                     1   
4                     0                   0                     1   

   Filipino Restaurant  French Restaurant  Fried Chicken Joint  \
0                    0                  0                    0   
1                    0

In [38]:
# add neighborhood column back to dataframe
nyc_onehot['Neighborhood'] = nyc_venues['Neighborhood'] 
nyc_onehot.head()

Afghan Restaurant  African Restaurant  American Restaurant  \
0                  0                   0                    0   
1                  0                   0                    0   
2                  0                   0                    0   
3                  0                   0                    0   
4                  0                   0                    0   

   Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
0                 0                       0                 0   
1                 0                       0                 0   
2                 0                       0                 0   
3                 0                       0                 0   
4                 0                       0                 0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Beer Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Belgian Restaurant  Bike Shop  Brazilian Restaurant  Burrito Place  \
0                   0          0                     0              0   
1                   0          0                     0              0   
2                   0          0                     0              0   
3                   0          0                     0              0   
4                   0          0                     0              0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                          0                     0                     1   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
0                     0                   0                     0   
1                     0                   0                     0   
2                     0                   1                     0   
3                     0                   0                     0   
4                     0                   0                     0   

   Cuban Restaurant  Dim Sum Restaurant  Dosa Place  Dumpling Restaurant  \
0                 0                   0           0                    0   
1                 0                   0           0                    0   
2                 0                   0           0                    0   
3                 0                   0           0                    0   
4                 0                   0           0                    0   

   Eastern European Restaurant  Empanada Restaurant  English Restaurant  \
0                            0                    0                   0   
1                            0                    0                   0   
2                            0                    0                   0   
3                            0                    0                   0   
4                            0                    0                   0   

   Ethiopian Restaurant  Falafel Restaurant  Fast Food Restaurant  \
0                     0                   0                     0   
1                     0                   0                     1   
2                     0                   0                     0   
3                     0                   0                     1   
4                     0                   0                     1   

   Filipino Restaurant  French Restaurant  Fried Chicken Joint  \
0                    0                  0                    0   
1                    0

In [39]:
# move neighborhood column to the first column
Neighborhood = nyc_onehot['Neighborhood']

nyc_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
nyc_onehot.insert(0, 'Neighborhood', Neighborhood)

nyc_onehot.head()

Neighborhood  Afghan Restaurant  African Restaurant  American Restaurant  \
0    Wakefield                  0                   0                    0   
1    Wakefield                  0                   0                    0   
2    Wakefield                  0                   0                    0   
3    Wakefield                  0                   0                    0   
4    Wakefield                  0                   0                    0   

   Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
0                 0                       0                 0   
1                 0                       0                 0   
2                 0                       0                 0   
3                 0                       0                 0   
4                 0                       0                 0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Beer Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Belgian Restaurant  Bike Shop  Brazilian Restaurant  Burrito Place  \
0                   0          0                     0              0   
1                   0          0                     0              0   
2                   0          0                     0              0   
3                   0          0                     0              0   
4                   0          0                     0              0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                          0                     0                     1   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
0                     0                   0                     0   
1                     0                   0                     0   
2                     0                   1                     0   
3                     0                   0                     0   
4                     0                   0                     0   

   Cuban Restaurant  Dim Sum Restaurant  Dosa Place  Dumpling Restaurant  \
0                 0                   0           0                    0   
1                 0                   0           0                    0   
2                 0                   0           0                    0   
3                 0                   0           0                    0   
4                 0                   0           0                    0   

   Eastern European Restaurant  Empanada Restaurant  English Restaurant  \
0                            0                    0                   0   
1                            0                    0                   0   
2                            0                    0                   0   
3                            0                    0                   0   
4                            0                    0                   0   

   Ethiopian Restaurant  Falafel Restaurant  Fast Food Restaurant  \
0                     0                   0                     0   
1                     0                   0                     1   
2                     0                   0                     0   
3                     0                   0                     1   
4                     0                   0                     1   

   Filipino Restaurant  French Restaurant  Fried Chicken Joint  \
0        

And let's examine the new dataframe size.

In [40]:
nyc_onehot.shape

(6726, 100)

Let's count venues of each category in each neighborhood

In [41]:
venue_counts = nyc_onehot.groupby('Neighborhood').sum()
venue_counts.head(5)

Afghan Restaurant  African Restaurant  American Restaurant  \
Neighborhood                                                                
Allerton                       0                   0                    1   
Annadale                       0                   0                    3   
Arden Heights                  0                   0                    3   
Arlington                      0                   0                    2   
Arrochar                       0                   0                    0   

               Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
Neighborhood                                                                
Allerton                      0                       0                 0   
Annadale                      0                       0                 0   
Arden Heights                 0                       0                 0   
Arlington                     0                       0                 1   
Arrochar                      0                       0                 0   

               Australian Restaurant  Austrian Restaurant  BBQ Joint  \
Neighborhood                                                           
Allerton                           0                    0          0   
Annadale                           0                    0          0   
Arden Heights                      0                    0          1   
Arlington                          0                    0          0   
Arrochar                           0                    0          0   

               Beer Store  Belgian Restaurant  Bike Shop  \
Neighborhood                                               
Allerton                0                   0          0   
Annadale                0                   0          0   
Arden Heights           0                   0          0   
Arlington               0                   0          0   
Arrochar                0                   0          0   

               Brazilian Restaurant  Burrito Place  Cajun / Creole Restaurant  \
Neighborhood                                                                    
Allerton                          0              0                          0   
Annadale                          0              0                          0   
Arden Heights                     0              0                          0   
Arlington                         0              0                          0   
Arrochar                          0              0                          0   

               Cantonese Restaurant  Caribbean Restaurant  \
Neighborhood                                                
Allerton                          0                     2   
Annadale                          0                     0   
Arden Heights                     0                     0   
Arlington                         0                     1   
Arrochar                          0                     0   

               Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
Neighborhood                                                                    
Allerton                          0                   3                     0   
Annadale                          0                   1                     0   
Arden Heights                     0                   2                     0   
Arlington                         0                   1                     0   
Arrochar                          0                   2                     0   

               Cuban Restaurant  Dim Sum Restaurant  Dosa Place  \
Neighborhood                                                      
Allerton                      0                   0           0   
Annadale                      0                   0           0   
Arden Heights                 0                   0           0   
Arlington                     0                   0           0   
Arrochar                      0            

Let's find out the top 10 food categories in NYC

In [42]:
venue_counts_described = venue_counts.describe().transpose()

In [44]:
venue_top10 = venue_counts_described.sort_values('max', ascending=False)[0:10]
venue_top10

count      mean       std  min  25%  50%  75%   max
Korean Restaurant     302.0  0.238411  1.352780  0.0  0.0  0.0  0.0  20.0
Caribbean Restaurant  302.0  1.238411  2.680578  0.0  0.0  0.0  1.0  16.0
Chinese Restaurant    302.0  2.129139  1.964776  0.0  1.0  2.0  3.0  15.0
Pizza Place           302.0  3.847682  2.134197  0.0  2.0  4.0  5.0  12.0
Indian Restaurant     302.0  0.377483  1.101142  0.0  0.0  0.0  0.0  12.0
Italian Restaurant    302.0  1.582781  1.814880  0.0  0.0  1.0  2.0  12.0
Fast Food Restaurant  302.0  2.125828  1.893527  0.0  1.0  2.0  3.0  10.0
Seafood Restaurant    302.0  0.476821  0.884225  0.0  0.0  0.0  1.0   8.0
Sushi Restaurant      302.0  0.562914  0.944128  0.0  0.0  0.0  1.0   6.0
Asian Restaurant      302.0  0.486755  0.861597  0.0  0.0  0.0  1.0   6.0

In [45]:
venue_top10 = venue_top10.index.values.tolist()

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [46]:
nyc_grouped = nyc_onehot.groupby('Neighborhood').mean().reset_index()
nyc_grouped.head()

Neighborhood  Afghan Restaurant  African Restaurant  American Restaurant  \
0       Allerton                0.0                 0.0             0.058824   
1       Annadale                0.0                 0.0             0.166667   
2  Arden Heights                0.0                 0.0             0.157895   
3      Arlington                0.0                 0.0             0.111111   
4       Arrochar                0.0                 0.0             0.000000   

   Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
0               0.0                     0.0          0.000000   
1               0.0                     0.0          0.000000   
2               0.0                     0.0          0.000000   
3               0.0                     0.0          0.055556   
4               0.0                     0.0          0.000000   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Beer Store  \
0                    0.0                  0.0   0.000000         0.0   
1                    0.0                  0.0   0.000000         0.0   
2                    0.0                  0.0   0.052632         0.0   
3                    0.0                  0.0   0.000000         0.0   
4                    0.0                  0.0   0.000000         0.0   

   Belgian Restaurant  Bike Shop  Brazilian Restaurant  Burrito Place  \
0                 0.0        0.0                   0.0            0.0   
1                 0.0        0.0                   0.0            0.0   
2                 0.0        0.0                   0.0            0.0   
3                 0.0        0.0                   0.0            0.0   
4                 0.0        0.0                   0.0            0.0   

   Cajun / Creole Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
0                        0.0                   0.0              0.117647   
1                        0.0                   0.0              0.000000   
2                        0.0                   0.0              0.000000   
3                        0.0                   0.0              0.055556   
4                        0.0                   0.0              0.000000   

   Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
0                   0.0            0.176471                   0.0   
1                   0.0            0.055556                   0.0   
2                   0.0            0.105263                   0.0   
3                   0.0            0.055556                   0.0   
4                   0.0            0.111111                   0.0   

   Cuban Restaurant  Dim Sum Restaurant  Dosa Place  Dumpling Restaurant  \
0               0.0                 0.0         0.0                  0.0   
1               0.0                 0.0         0.0                  0.0   
2               0.0                 0.0         0.0                  0.0   
3               0.0                 0.0         0.0                  0.0   
4               0.0                 0.0         0.0                  0.0   

   Eastern European Restaurant  Empanada Restaurant  English Restaurant  \
0                          0.0                  0.0                 0.0   
1                          0.0                  0.0                 0.0   
2                          0.0                  0.0                 0.0   
3                          0.0                  0.0                 0.0   
4                          0.0                  0.0                 0.0   

   Ethiopian Restaurant  Falafel Restaurant  Fast Food Restaurant  \
0                   0.0                 0.0              0.235294   
1                   0.0                 0.0              0.000000   
2                   0.0                 0.0              0.000000   
3                   0.0                 0.0              0.111111   
4                   0.0                 0.0              0.000000   

   Filipino Restaurant  French Restaurant  Fried Chicken Joint  \

Let's confirm the new size

In [47]:
nyc_grouped.shape

(302, 100)

Let's write a function to sort the venues in descending order.

In [48]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 5 venues for each neighborhood.

In [51]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))  

In [52]:
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = nyc_grouped['Neighborhood']


In [53]:
for ind in np.arange(nyc_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(nyc_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Allerton           Pizza Place  Fast Food Restaurant   
1       Annadale           Pizza Place      Sushi Restaurant   
2  Arden Heights           Pizza Place   American Restaurant   
3      Arlington           Pizza Place    Spanish Restaurant   
4       Arrochar           Pizza Place    Italian Restaurant   

  3rd Most Common Venue      4th Most Common Venue     5th Most Common Venue  
0    Chinese Restaurant       Caribbean Restaurant       Fried Chicken Joint  
1   American Restaurant         Italian Restaurant       Japanese Restaurant  
2    Italian Restaurant           Sushi Restaurant        Mexican Restaurant  
3   Peruvian Restaurant        American Restaurant      Fast Food Restaurant  
4    Chinese Restaurant  Middle Eastern Restaurant  Mediterranean Restaurant

List the top 40 to see where African Restaurants fit in

In [54]:
venue_top40 = venue_counts_described.sort_values('max', ascending=False)[0:40]
venue_top40

count      mean       std  min  25%  50%  \
Korean Restaurant                302.0  0.238411  1.352780  0.0  0.0  0.0   
Caribbean Restaurant             302.0  1.238411  2.680578  0.0  0.0  0.0   
Chinese Restaurant               302.0  2.129139  1.964776  0.0  1.0  2.0   
Pizza Place                      302.0  3.847682  2.134197  0.0  2.0  4.0   
Indian Restaurant                302.0  0.377483  1.101142  0.0  0.0  0.0   
Italian Restaurant               302.0  1.582781  1.814880  0.0  0.0  1.0   
Fast Food Restaurant             302.0  2.125828  1.893527  0.0  1.0  2.0   
Seafood Restaurant               302.0  0.476821  0.884225  0.0  0.0  0.0   
Sushi Restaurant                 302.0  0.562914  0.944128  0.0  0.0  0.0   
Asian Restaurant                 302.0  0.486755  0.861597  0.0  0.0  0.0   
Spanish Restaurant               302.0  0.483444  0.895573  0.0  0.0  0.0   
Fried Chicken Joint              302.0  1.152318  1.398853  0.0  0.0  1.0   
Thai Restaurant                  302.0  0.367550  0.819614  0.0  0.0  0.0   
Latin American Restaurant        302.0  0.413907  0.809539  0.0  0.0  0.0   
Mexican Restaurant               302.0  1.125828  1.168835  0.0  0.0  1.0   
Eastern European Restaurant      302.0  0.059603  0.377644  0.0  0.0  0.0   
American Restaurant              302.0  1.056291  1.050301  0.0  0.0  1.0   
Southern / Soul Food Restaurant  302.0  0.211921  0.566029  0.0  0.0  0.0   
Vietnamese Restaurant            302.0  0.145695  0.459631  0.0  0.0  0.0   
Filipino Restaurant              302.0  0.092715  0.388386  0.0  0.0  0.0   
New American Restaurant          302.0  0.195364  0.513840  0.0  0.0  0.0   
Greek Restaurant                 302.0  0.182119  0.499114  0.0  0.0  0.0   
Japanese Restaurant              302.0  0.440397  0.716273  0.0  0.0  0.0   
Turkish Restaurant               302.0  0.096026  0.383268  0.0  0.0  0.0   
Vegetarian / Vegan Restaurant    302.0  0.168874  0.447909  0.0  0.0  0.0   
South American Restaurant        302.0  0.096026  0.365521  0.0  0.0  0.0   
Sri Lankan Restaurant            302.0  0.046358  0.290211  0.0  0.0  0.0   
BBQ Joint                        302.0  0.337748  0.569285  0.0  0.0  0.0   
Russian Restaurant               302.0  0.026490  0.197897  0.0  0.0  0.0   
Middle Eastern Restaurant        302.0  0.188742  0.489904  0.0  0.0  0.0   
Mediterranean Restaurant         302.0  0.158940  0.432757  0.0  0.0  0.0   
Malay Restaurant                 302.0  0.036424  0.220233  0.0  0.0  0.0   
Steakhouse                       302.0  0.172185  0.419802  0.0  0.0  0.0   
Taco Place                       302.0  0.291391  0.541800  0.0  0.0  0.0   
Kosher Restaurant                302.0  0.049669  0.259407  0.0  0.0  0.0   
Tapas Restaurant                 302.0  0.092715  0.323005  0.0  0.0  0.0   
African Restaurant               302.0  0.096026  0.316832  0.0  0.0  0.0   
Peruvian Restaurant              302.0  0.102649  0.314742  0.0  0.0  0.0   
Afghan Restaurant                302.0  0.023179  0.171351  0.0  0.0  0.0   
Cantonese Restaurant             302.0  0.036424  0.220233  0.0  0.0  0.0   

                                 75%   max  
Korean Restaurant                0.0  20.0  
Caribbean Restaurant             1.0  16.0  
Chinese Restaurant               3.0  15.0  
Pizza Place                      5.0  12.0  
Indian Restaurant                0.0  12.0  
Italian Restaurant               2.0  12.0  
Fast Food Restaurant             3.0  10.0  
Seafood Restaurant               1.0   8.0  
Sushi Restaurant                 1.0   6.0  
Asian Restaurant                 1.0   6.0  
Spanish Restaurant               1.0   6.0  
Fried Chicken Joint              2.0   6.0  
Thai Restaurant                  0.0   6.0  
Latin American Restaurant        1.0   5.0  
Mexican Restaurant               2.0   5.0  
Eastern European Restaurant      0.0   5.0  
American Restaurant              2.0   5.0  
Southern / Soul Food Restaurant  0.0   4.0  
Vietnamese Restauran

###### Conclusion

Based on the analysis performed and exploratory data analysis, there seems to insufficient interest in African themed food and as such no further analysis will be done to eplore teh data further.
Due to the limited occurrence of African Restaurants, it will not be visualised either.